src/build_helper/src/git.rs

```diff
@@ -27,145 +28,195 @@ pub fn output_result(cmd: &mut Command) -> Result<String, String> {
-    if let Some(git_dir) = git_dir {
-        git.current_dir(git_dir);
+    git.current_dir(git_dir);
+
+    git.args(["rev-list", "-n1", "--first-parent", "HEAD", "--author", author]);
+
+    if !target_paths.is_empty() {
+        git.arg("--").args(target_paths);
     }
-    Ok(output_result(git.arg("merge-base").arg(&updated_master).arg("HEAD"))?.trim().to_owned())
+    let output = output_result(&mut git)?.trim().to_owned();
+    if output.is_empty() { Ok(None) } else { Ok(Some(output)) }
 }
 
-/// Searches for the nearest merge commit in the repository that also exists upstream.
+/// Returns the most recent commit found in the local history that should definitely
+/// exist upstream. We identify upstream commits by the e-mail of the commit author.
 ///
-/// It looks for the most recent commit made by the merge bot by matching the author's email
-/// address with the merge bot's email.
-pub fn get_closest_merge_commit(
+/// If `include_head` is false, the HEAD (current) commit will be ignored and only
+/// its parents will be searched. This is useful for try/auto CI, where HEAD is
+/// actually a commit made by bors, although it is not upstream yet.
+fn get_closest_upstream_commit(
     git_dir: Option<&Path>,
     config: &GitConfig<'_>,
-    target_paths: &[&str],
-) -> Result<String, String> {
+    env: CiEnv,
+) -> Result<Option<String>, String> {
     let mut git = Command::new("git");
 
     if let Some(git_dir) = git_dir {
         git.current_dir(git_dir);
     }
 
-    let channel = include_str!("../../ci/channel").trim();
-
-    let merge_base = {
-        if CiEnv::is_ci() &&
-            // FIXME: When running on rust-lang managed CI and it's not a nightly build,
-            // `git_upstream_merge_base` fails with an error message similar to this:
-            // ```
-            //    called `Result::unwrap()` on an `Err` value: "command did not execute successfully:
-            //    cd \"/checkout\" && \"git\" \"merge-base\" \"origin/master\" \"HEAD\"\nexpected success, got: exit status: 1\n"
-            // ```
-            // Investigate and resolve this issue instead of skipping it like this.
-            (channel == "nightly" || !CiEnv::is_rust_lang_managed_ci_job())
-        {
-            git_upstream_merge_base(config, git_dir).unwrap()
-        } else {
-            // For non-CI environments, ignore rust-lang/rust upstream as it usually gets
-            // outdated very quickly.
-            "HEAD".to_string()
+    let base = match env {
+        CiEnv::None => "HEAD",
+        CiEnv::GitHubActions => {
+            // On CI, we always have a merge commit at the tip.
+            // We thus skip it, because although it can be created by
+            // `config.git_merge_commit_email`, it should not be upstream.
+            "HEAD^1"
         }
     };
-
     git.args([
         "rev-list",
         &format!("--author={}", config.git_merge_commit_email),
         "-n1",
         "--first-parent",
```

Comment:
Is first-parent right here? IIRC that was causing Ralf some issues in a recent PR?

Reply:
No, the original problem was exactly the `--first-parent`. See https://github.com/rust-lang/rust/issues/101907#issuecomment-2677867833 and the following comments.

Reply:
Thank you, now I understand. The most recent commit should fix this.